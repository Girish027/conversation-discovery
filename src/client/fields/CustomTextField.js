import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  TextField
} from '@tfs/ui-components';
import ObjectUtils from '../utils/ObjectUtils';
import { customTextField } from '../styles/customComponentsStyles';
import ValidationPlaceholder from './ValidationPlaceholder';

export default class CustomTextField extends Component {
  constructor(props) {
    super(props);
    this.schema = props.schema;
    const value = ObjectUtils.isEmptyOrNull(props.formData)
      ? this.schema.default
      : props.formData;
    this.state = {
      value,
      validationMessage: ''
    };

    this.handleChange = this.handleChange.bind(this);
  }

  getLabel() {
    return (
      <div
        style={customTextField.label}
      >
        {this.schema.title}
      </div>
    );
  }

  getField() {
    const { validationMessage } = this.state;
    let validationMsg = '';

    if (validationMessage) {
      validationMsg = validationMessage.slice(0, validationMessage.indexOf('('));
    }

    return (
      <React.Fragment>
        <TextField
          name={this.props.name}
          onChange={this.handleChange}
          defaultValue={this.state.value}
          placeholder={this.schema.placeholder}
          styleOverride={customTextField.field}
          minLength={this.schema.minLength}
          maxLength={this.schema.maxLength}
        />
        {validationMessage && (
          <ValidationPlaceholder validationMessage={validationMsg} />
        )}
      </React.Fragment>
    );
  }

  handleChange(event) {
    const { value, validationMessage } = event.target;
    this.setState({ value, validationMessage });
    this.props.onChange(value);
  }

  render() {
    return (
      <div
        style={customTextField}
      >
        {this.getLabel()}
        {this.getField()}
      </div>
    );
  }
}

CustomTextField.propTypes = {
  name: PropTypes.string,
  schema: PropTypes.object.isRequired,
  formData: PropTypes.string,
  onChange: PropTypes.func
};

CustomTextField.defaultProps = {
  onChange: () => {},
  name: 'textfield',
};
