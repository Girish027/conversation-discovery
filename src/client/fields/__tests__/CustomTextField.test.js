import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CustomTextField from '../CustomTextField';

describe('<CustomTextField />', function () {
  let wrapper;

  const props = {
    name: 'testName',
    schema: {
      maxLength: 150,
      minLength: 5,
      placeholder: 'Name this iteration',
      title: 'Name',
      type: 'string',
    },
    uiSchema: {},
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = shallow(<CustomTextField
        {...props}
      />);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly for default props', () => {
      wrapper = shallow(<CustomTextField
        {...props}
      />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    beforeEach(function () {
      props.onChange = jest.fn();
    });

    describe('onChange', function () {
      test('should pass the props and change the state with new value in the form', () => {
        wrapper = shallow(<CustomTextField
          {...props}
        />);

        wrapper.instance().handleChange({
          target: {
            name: 'runName',
            value: 'Run Name'
          }
        });
        expect(props.onChange).toHaveBeenCalledWith('Run Name');
        expect(wrapper.state().value).toEqual('Run Name');
      });

      test('should chamge the state in case of validation failure', () => {
        wrapper = shallow(<CustomTextField
          {...props}
        />);

        expect(wrapper.state().validationMessage).toEqual('');
        wrapper.instance().handleChange({
          target: {
            name: 'runName',
            value: 'Run Name',
            validationMessage: 'Name should not be less than 4 char'
          }
        });
        expect(wrapper.state().validationMessage).toEqual('Name should not be less than 4 char');
      });
    });
  });
});
