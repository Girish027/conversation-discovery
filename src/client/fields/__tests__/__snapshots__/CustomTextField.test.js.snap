// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomTextField /> Snapshots renders correctly for default props 1`] = `
<div
  style={
    Object {
      "field": Object {
        "background": "#FFFFFF",
        "border": "0px solid #DCE1E8",
        "borderRadius": "2px",
        "input": Object {
          "fontSize": "14px",
          "letterSpacing": 0,
          "textAlign": "left",
        },
        "width": "99%",
      },
      "label": Object {
        "color": "#6E6E6E",
        "fontSize": "14px",
        "letterSpacing": 0,
        "paddingBottom": "5px",
        "textAlign": "left",
      },
      "marginTop": "20px",
    }
  }
>
  <div
    style={
      Object {
        "color": "#6E6E6E",
        "fontSize": "14px",
        "letterSpacing": 0,
        "paddingBottom": "5px",
        "textAlign": "left",
      }
    }
  >
    Name
  </div>
  <withTheme(TextField)
    maxLength={150}
    minLength={5}
    name="testName"
    onChange={[Function]}
    placeholder="Name this iteration"
    styleOverride={
      Object {
        "background": "#FFFFFF",
        "border": "0px solid #DCE1E8",
        "borderRadius": "2px",
        "input": Object {
          "fontSize": "14px",
          "letterSpacing": 0,
          "textAlign": "left",
        },
        "width": "99%",
      }
    }
  />
</div>
`;
